lib/chef/provider/package.rb

```diff
@@ -412,59 +412,60 @@ def target_version_array
-                  logger.trace("#{new_resource} #{package_name} has no candidate_version to upgrade to")
-                  target_version_array.push(nil)
-                elsif current_version.nil?
-                  logger.trace("#{new_resource} has no existing installed version. Installing install #{candidate_version}")
-                  target_version_array.push(candidate_version)
-                elsif !allow_downgrade && version_compare(current_version, candidate_version) == 1
-                  logger.trace("#{new_resource} #{package_name} has installed version #{current_version}, which is newer than available version #{candidate_version}. Skipping...)")
-                  target_version_array.push(nil)
-                else
-                  logger.trace("#{new_resource} #{package_name} is out of date, will upgrade to #{candidate_version}")
-                  target_version_array.push(candidate_version)
-                end
-
-              when :install
-                if new_version
-                  if version_requirement_satisfied?(current_version, new_version)
-                    logger.trace("#{new_resource} #{package_name} #{current_version} satisfies #{new_version} requirement")
-                    target_version_array.push(nil)
-                  elsif current_version && !allow_downgrade && version_compare(current_version, new_version) == 1
-                    logger.warn("#{new_resource} #{package_name} has installed version #{current_version}, which is newer than available version #{new_version}. Skipping...)")
-                    target_version_array.push(nil)
-                  else
-                    logger.trace("#{new_resource} #{package_name} #{current_version} needs updating to #{new_version}")
-                    target_version_array.push(new_version)
-                  end
-                elsif current_version.nil?
-                  logger.trace("#{new_resource} #{package_name} not installed, installing #{candidate_version}")
-                  target_version_array.push(candidate_version)
-                else
-                  logger.trace("#{new_resource} #{package_name} #{current_version} already installed")
-                  target_version_array.push(nil)
-                end
-
-              else
-                # in specs please test the public interface provider.run_action(:install) instead of provider.action_install
-                raise "internal error - target_version_array in package provider does not understand this action"
-              end
+              target = target_version(package_name, new_version, current_version, candidate_version)
+              target_version_array.push(target == current_version ? nil : target)
             end
-
             target_version_array
           end
       end
 
+      # Returns the version a package should have after the install/upgrade operation is complete.
+      #
+      # @return [String] target package version
```

Comment:
now that you've separated this out the `nil` case(s) look like a bug. It's worth adding documentation here that this returns nil if there is nothing to do.

Reply:
I am not sure I am understanding what you are referring to.  I (also?) thought it would be confusing to have `nil` as a target version, so the `target_version` method never returns `nil`.  The cases that do not immediately return a new version of the package to install/update fall through to the end of the function and return `current_version`.  Since `current_version.nil?` is handled for each value of `action`, `current_version` should never be `nil` when it is returned.

The `target_version_array` method translates a target version of `current_version` to `nil` when constructing its array, and the meaning of `nil` in that case is documented in the description of `target_version_array`.

Reply:
oh I missed the final return here.

But that ... won't work. I think you break :upgrade here.

`nil` in the case of multipackage means 'latest' for :upgrade (ie do work if there's a newer version in the repo), or "anything" for :install (ie do nothing if anything is installed). For the :upgrade path, you'll likely need to return `nil` specifically.

Reply:
I do not understand to what you are referring.  `nil` in the `target_version_array` means "the package is not being modified" for both the `:upgrade` and `:install` cases (per line 407).  In all of the cases where `nil` was being pushed onto the stack in the `target_version_array` method, the `target_version` method returns `current_version`, meaning the package is expected to be at the same version as is currently installed.  The `target_version_array` method now simply retrieves the `target_version` for each package, and maps `current_version` responses back to the `nil`s it uses to represent a no-op on the package in its array.

Are we perhaps talking about the use of `nil` in different parts of the code?